clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -290,18 +317,47 @@ public void poll(long timeout) throws IOException {
 
         clear();
 
-        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty())
+        boolean dataInBuffers = !keysWithBytesBuffered.isEmpty();
+        
+        if (hasStagedReceives() || !immediatelyConnectedKeys.isEmpty() || (madeProgressLastPoll && dataInBuffers))
             timeout = 0;
 
+        if (!memoryPool.isOutOfMemory() && (outOfMemory || dataInBuffers)) {
+            //server has recovered from memory pressure. unmute any channel not explicitly muted for other reasons
+            log.trace("Broker no longer low on memory - unmuting incoming sockets");
+            for (KafkaChannel channel : channels.values()) {
+                if (channel.isInMutableState() && !explicitlyMutedChannels.contains(channel)) {
+                    channel.unmute();
+                }
+            }
+            outOfMemory = false;
+        }
+
         /* check ready keys */
         long startSelect = time.nanoseconds();
         int readyKeys = select(timeout);
         long endSelect = time.nanoseconds();
         this.sensors.selectTime.record(endSelect - startSelect, time.milliseconds());
 
-        if (readyKeys > 0 || !immediatelyConnectedKeys.isEmpty()) {
-            pollSelectionKeys(this.nioSelector.selectedKeys(), false, endSelect);
-            pollSelectionKeys(immediatelyConnectedKeys, true, endSelect);
+        if (readyKeys > 0 || !immediatelyConnectedKeys.isEmpty() || dataInBuffers) {
+            Set<SelectionKey> keysWithBytesFromSocket = this.nioSelector.selectedKeys();
+            keysWithBytesBuffered.removeAll(keysWithBytesFromSocket); //so no channel gets polled twice
+
+            boolean madeProgress = false;
+
+            //poll from channels that have buffered data (but nothing more from the underlying socket)
+            if (dataInBuffers) {
+                Set<SelectionKey> toPoll = keysWithBytesBuffered;
+                keysWithBytesBuffered = new HashSet<>(); //poll() calls will repopulate if needed
+                madeProgress |= pollSelectionKeys(toPoll, false, endSelect);
+            }
+            //poll from channels where the underlying socket has more data
+            madeProgress |= pollSelectionKeys(keysWithBytesFromSocket, false, endSelect);
+            madeProgress |= pollSelectionKeys(immediatelyConnectedKeys, true, endSelect);
+
+            madeProgressLastPoll = madeProgress;
+        } else {
+            madeProgressLastPoll = true; //no work is also "progress"
         }
 
         addToCompletedReceives();
```

Comment:
The current implementation of `addToCompletedReceives` moves receives from staged to completed state if the channel is not muted. I think it will better to replace `!channel.isMute()` with  `!explicitlyMutedChannels.contains(channel)`. Buffers have already been allocated for the staged receives, so we should allow them to make progress and release the buffers.